### PR TITLE
Cardano implementation of Ledger Tables

### PR DESCRIPTION
**Pull request by Jasagredo**
# Description

This PR contains the changes required in the non-test packages of the Cardano bundle in order to support Ledger Tables.

Part of #4212

# Checklist

- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Commits have useful messages
    - [ ] The documentation has been properly updated
    - [ ] New tests are added if needed and existing tests are updated
    - [ ] Any changes affecting Consensus packages must have an entry in the appropriate `changelog.d` directory created using [`scriv`](https://github.com/input-output-hk/scriv). If in doubt, see the [Consensus release process](../ouroboros-consensus/docs/ReleaseProcess.md).
    - [ ] If this branch changes Network and has any consequences for downstream repositories or end users, said changes must be documented in [`interface-CHANGELOG.md`](../docs/interface-CHANGELOG.md)
    - [ ] If serialization changes, user-facing consequences (e.g. replay from genesis) are confirmed to be intentional.
- Pull Request
    - [x] Self-reviewed the diff
    - [x] Useful pull request description at least containing the following information:
      - What does this PR change?
      - Why these changes were needed?
      - How does this affect downstream repositories and/or end-users?
      - Which ticket does this PR close (if any)? If it does, is it [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)?
    - [x] Reviewer requested

### CODE REVIEW

**Comment by Jasagredo** [comment]
Current dependencies on/for this PR:
* master
  * **PR #4352** <a href="https://app.graphite.dev/github/pr/input-output-hk/ouroboros-network/4352" target="_blank">Graphite</a> 
    * **PR #4341** <a href="https://app.graphite.dev/github/pr/input-output-hk/ouroboros-network/4341" target="_blank">Graphite</a> 
      * **PR #4342** <a href="https://app.graphite.dev/github/pr/input-output-hk/ouroboros-network/4342" target="_blank">Graphite</a> 
        * **PR #4343** <a href="https://app.graphite.dev/github/pr/input-output-hk/ouroboros-network/4343" target="_blank">Graphite</a> 
          * **PR #4344** <a href="https://app.graphite.dev/github/pr/input-output-hk/ouroboros-network/4344" target="_blank">Graphite</a> 
            * **PR #4366** <a href="https://app.graphite.dev/github/pr/input-output-hk/ouroboros-network/4366" target="_blank">Graphite</a> 
              * **PR #4363** <a href="https://app.graphite.dev/github/pr/input-output-hk/ouroboros-network/4363" target="_blank">Graphite</a> 
                * **PR #4391** <a href="https://app.graphite.dev/github/pr/input-output-hk/ouroboros-network/4391" target="_blank">Graphite</a>  👈
    * **PR #4316** <a href="https://app.graphite.dev/github/pr/input-output-hk/ouroboros-network/4316" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/input-output-hk/ouroboros-network/4391?utm_source=stack-comment).

**Comment by Jasagredo** [comment]
Current dependencies on/for this PR:
* master
  * **PR #4352** <a href="https://app.graphite.dev/github/pr/input-output-hk/ouroboros-network/4352" target="_blank">Graphite</a> 
    * **PR #4341** <a href="https://app.graphite.dev/github/pr/input-output-hk/ouroboros-network/4341" target="_blank">Graphite</a> 
      * **PR #4342** <a href="https://app.graphite.dev/github/pr/input-output-hk/ouroboros-network/4342" target="_blank">Graphite</a> 
        * **PR #4343** <a href="https://app.graphite.dev/github/pr/input-output-hk/ouroboros-network/4343" target="_blank">Graphite</a> 
          * **PR #4344** <a href="https://app.graphite.dev/github/pr/input-output-hk/ouroboros-network/4344" target="_blank">Graphite</a> 
            * **PR #4366** <a href="https://app.graphite.dev/github/pr/input-output-hk/ouroboros-network/4366" target="_blank">Graphite</a> 
              * **PR #4363** <a href="https://app.graphite.dev/github/pr/input-output-hk/ouroboros-network/4363" target="_blank">Graphite</a> 
                * **PR #4391** <a href="https://app.graphite.dev/github/pr/input-output-hk/ouroboros-network/4391" target="_blank">Graphite</a>  👈
    * **PR #4316** <a href="https://app.graphite.dev/github/pr/input-output-hk/ouroboros-network/4316" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/input-output-hk/ouroboros-network/4391?utm_source=stack-comment).

File: ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/CanHardFork.hs [hunk 1]
```diff
@@ -350,6 +380,310 @@ instance CardanoHardForkConstraints c => CanHardFork (CardanoEras c) where
...
+          InjectLedgerTables
+           { applyInjectLedgerTables = \(ShelleyLedgerTables lt) -> CardanoLedgerTables $ mapMK f lt
+           , applyDistribLedgerTables = \(CardanoLedgerTables lt) -> ShelleyLedgerTables $ mapMK g lt
+           }
+        where
+          f :: Core.TxOut era -> CardanoTxOut c
+          f = ShelleyTxOut . injectNS idx . TxOutWrapper
+
+          g :: CardanoTxOut c -> Core.TxOut era
+          g = undefined --ShelleyTxOut . injectNS idx . TxOutWrapper
```
Comment (Jasagredo) [inline ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/CanHardFork.hs:684]: This is still undefined!!

File: ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/CanHardFork.hs [hunk 2]
```diff
@@ -348,6 +378,307 @@ instance CardanoHardForkConstraints c => CanHardFork (CardanoEras c) where
...
+  zipLedgerTablesA     f                         (CardanoLedgerTables l) (CardanoLedgerTables r) = CardanoLedgerTables <$> f l r
+  zipLedgerTables2A    f (CardanoLedgerTables l) (CardanoLedgerTables c) (CardanoLedgerTables r) = CardanoLedgerTables <$> f l c r
+  foldLedgerTables     f                                                 (CardanoLedgerTables x) = f x
+  foldLedgerTables2    f                         (CardanoLedgerTables l) (CardanoLedgerTables r) = f l r
+
+  namesLedgerTables = CardanoLedgerTables { cardanoUTxOTable = NameMK "cardanoUTxOTable" }
+
+instance CardanoHardForkConstraints c
+      => CanSerializeLedgerTables (LedgerState (CardanoBlock c)) where
+    codecLedgerTables = undefined
```
Comment (Jasagredo) [inline ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/CanHardFork.hs:567]: We cannot currently define this. We will have to find out a different solution next week.

File: ouroboros-consensus-cardano/README.md [hunk 3]
```diff
@@ -3,3 +3,7 @@
 This package contains:
 
 * `src` : Support for Hard Fork between Cardano eras.
+
+* `tools/ledger-db-backends-checker`: simple tool (for the UTxO-HD feature) to
+  compare the contents of an in-memory backing store and an LMDB backing store
+  (i.e., it checks if they have the same slot number and values).
```
Comment (jorisdral) [inline ouroboros-consensus-cardano/README.md:9]: Remove if https://github.com/input-output-hk/ouroboros-network/pull/4391/files?diff=unified&w=0#r1133748533 is done

File: ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Tables.hs [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+                  withLT innerSt
+                    $ applyDistribLedgerTables
+                        (projectNP i hardForkInjectLedgerTables)
+                        tbs
+              }
+
+-- Note that this is a HardForkBlock instance, but it's not compositional. This
+-- is because the LedgerTables relies on knowledge specific to Cardano and we
+-- have so far not found a pleasant way to express that compositionally.
+instance CardanoHardForkConstraints c => HasLedgerTables (LedgerState (CardanoBlock c)) where
```
Comment (jorisdral) [inline ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Tables.hs:136]: Wrap line

File: ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Tables.hs [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+
+  pureLedgerTables     f                                                                         = CardanoLedgerTables f
+  mapLedgerTables      f                                                 (CardanoLedgerTables x) = CardanoLedgerTables (f x)
+  traverseLedgerTables f                                                 (CardanoLedgerTables x) = CardanoLedgerTables <$> f x
+  zipLedgerTables      f                         (CardanoLedgerTables l) (CardanoLedgerTables r) = CardanoLedgerTables (f l r)
+  zipLedgerTables2     f (CardanoLedgerTables l) (CardanoLedgerTables c) (CardanoLedgerTables r) = CardanoLedgerTables (f l c r)
+  zipLedgerTablesA     f                         (CardanoLedgerTables l) (CardanoLedgerTables r) = CardanoLedgerTables <$> f l r
+  zipLedgerTables2A    f (CardanoLedgerTables l) (CardanoLedgerTables c) (CardanoLedgerTables r) = CardanoLedgerTables <$> f l c r
+  foldLedgerTables     f                                                 (CardanoLedgerTables x) = f x
+  foldLedgerTables2    f                         (CardanoLedgerTables l) (CardanoLedgerTables r) = f l r
```
Comment (jorisdral) [inline ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Tables.hs:162]: Wrap lines

File: ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Tables.hs [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+                                             (Core.toEraCBOR @(ShelleyEra c))
+                                             toCBOR
+                                             (Core.fromEraCBOR @(ShelleyEra c))
+                                             fromCBOR)
+
+{-------------------------------------------------------------------------------
+  LedgerTablesCanHardFork
+-------------------------------------------------------------------------------}
+
+instance CardanoHardForkConstraints c => LedgerTablesCanHardFork (CardanoEras c) where
```
Comment (jorisdral) [inline ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Tables.hs:222]: Wrap lines

File: ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/CanHardFork.hs [hunk 7]
```diff
@@ -392,25 +414,33 @@ translateLedgerStateByronToShelleyWrapper ::
...
+                      (byronLedgerTipBlockNo ledgerByron)
+                  , shelleyLedgerState =
+                      SL.translateToShelleyLedgerState
+                        (toFromByronTranslationContext (shelleyLedgerGenesis cfgShelley))
+                        epochNo
+                        (byronLedgerState ledgerByron)
+                  , shelleyLedgerTransition =
+                      ShelleyTransitionInfo{shelleyAfterVoting = 0}
+                  , shelleyLedgerTables = emptyLedgerTables
+                }
```
Comment (jorisdral) [inline ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/CanHardFork.hs:441]:
```suggestion
                  }
```

File: ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/CanHardFork.hs [hunk 8]
```diff
@@ -517,13 +547,62 @@ translateLedgerStateShelleyToAllegraWrapper ::
...
+                  -- thin air and when the time comes in which ticking produces
+                  -- differences, we will have to revisit this.
+                  avvmsAsDeletions = ShelleyLedgerTables
+                                   . DiffMK
+                                   . Diff.Internal.Diff
+                                   . Map.map (  Diff.Internal.singletonDelete
+                                              . unTxOutWrapper
+                                              . SL.translateEra' ()
+                                              . TxOutWrapper
+                                             )
```
Comment (jorisdral) [inline ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/CanHardFork.hs:581]: Consider using `fromMapDeletes` from `diff-containers`

File: ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/CanHardFork.hs [hunk 9]
```diff
@@ -517,13 +547,62 @@ translateLedgerStateShelleyToAllegraWrapper ::
      (PraosCrypto c, DSignable c (Hash c EraIndependentTxBody))
   => RequiringBoth
        WrapLedgerConfig
-       (Translate LedgerState)
+       TranslateLedgerState
        (ShelleyBlock (TPraos c) (ShelleyEra c))
        (ShelleyBlock (TPraos c) (AllegraEra c))
 translateLedgerStateShelleyToAllegraWrapper =
```
Comment (jorisdral) [inline ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/CanHardFork.hs:553]: Wrap lines below

File: ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/CanHardFork.hs [hunk 10]
```diff
@@ -392,25 +414,33 @@ translateLedgerStateByronToShelleyWrapper ::
      )
   => RequiringBoth
        WrapLedgerConfig
-       (Translate LedgerState)
+       TranslateLedgerState
        ByronBlock
        (ShelleyBlock (TPraos c) (ShelleyEra c))
 translateLedgerStateByronToShelleyWrapper =
```
Comment (jorisdral) [inline ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/CanHardFork.hs:420]: The documentation of `translateLedgerStateWith` in the `TranslateLedgerState` datatype says that there are two cases where it is important that `translateLedgerStateWith` returns `DiffMK`. However, in this translation, the ledger tables are empty! Is this comment outdated?

File: ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/CanHardFork.hs [hunk 11]
```diff
@@ -549,13 +628,24 @@ translateLedgerStateAllegraToMaryWrapper ::
      (PraosCrypto c, DSignable c (Hash c EraIndependentTxBody))
```
Comment (jorisdral) [inline ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/CanHardFork.hs:628]: The section header above is wrong: it should say "Translation from Allegra to Mary"

File: ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/CanHardFork.hs [hunk 12]
```diff
@@ -549,13 +628,24 @@ translateLedgerStateAllegraToMaryWrapper ::
...
+              . Comp
+              . Flip
+        , translateLedgerTablesWith =
+            \ShelleyLedgerTables { shelleyUTxOTable = diffMK } ->
+             ShelleyLedgerTables { shelleyUTxOTable = fmap (SL.translateEra' ()) diffMK
+                                 }
+        }
 
 {-------------------------------------------------------------------------------
   Translation from Allegra to Mary
```
Comment (jorisdral) [inline ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/CanHardFork.hs:651]: And this section header is redundant

**Review by jorisdral — COMMENTED** [review]
Looks good -- I have  some comments, most of which are formatting/style/documentation related

File: ouroboros-consensus-cardano/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (dnadales) [reply ouroboros-consensus-cardano/README.md:9]: :100:

[the remark this answers is not in the document]
Comment (dnadales) [reply ouroboros-consensus-cardano/README.md:9]: Should we create an issue for this?

File: ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Tables.hs [hunk 13]
```diff
@@ -0,0 +1,327 @@
...
+import           Ouroboros.Consensus.Cardano.Block
+import           Ouroboros.Consensus.Cardano.CanHardFork
+import qualified Ouroboros.Consensus.Protocol.Praos as Praos
+import qualified Ouroboros.Consensus.Protocol.TPraos as TPraos
+
+{-------------------------------------------------------------------------------
+  Ledger Tables
+-------------------------------------------------------------------------------}
+
+-- We reuse this for both HasLedgerTables and HasTickedLedgerTables instances,
```
Comment (dnadales) [inline ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Tables.hs:76]: I wonder if this is missing:

```suggestion
-- | We reuse this for both HasLedgerTables and HasTickedLedgerTables instances,
```

File: ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Tables.hs [hunk 14]
```diff
@@ -0,0 +1,327 @@
...
+-- We can't reuse 'Translate' because we don't have the 'EpochNo' it requires.
+newtype TranslateTxOutWrapper era1 era2 =
+    TranslateTxOutWrapper (Core.TxOut era1 -> Core.TxOut era2)
+
+-- | Auxiliary @SOP@ combinator
+--
+-- WARNING: The functions in the result fail if the 'NS' argument's tag
+-- precedes the 'NP' index, i.e. if we try to translate into previous eras.
+--
+-- TODO use an accumulator instead of this quadratic traversal
```
Comment (dnadales) [inline ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Tables.hs:295]: https://github.com/input-output-hk/ouroboros-network/issues/4463

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Ledger.hs [hunk 15]
```diff
@@ -232,27 +243,205 @@ newtype ShelleyTransition = ShelleyTransitionInfo {
...
+  pureLedgerTables f = ShelleyLedgerTables f
+
+  mapLedgerTables f (ShelleyLedgerTables utxo) = ShelleyLedgerTables (f utxo)
+
+  traverseLedgerTables f (ShelleyLedgerTables utxo) = ShelleyLedgerTables <$> f utxo
+
+  zipLedgerTables f (ShelleyLedgerTables utxoL) (ShelleyLedgerTables utxoR) =
+      ShelleyLedgerTables (f utxoL utxoR)
+
+  zipLedgerTables2
```
[the remark this answers is not in the document]
Comment (dnadales) [reply ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Ledger.hs:281]: We are zipping 3 tables indeed. I like your suggestion.

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Mempool.hs [hunk 16]
```diff
@@ -237,33 +251,43 @@ applyShelleyTx cfg wti slot (ShelleyTx _ tx) st = do
...
+        st2 = set theLedgerLens mempoolState' st1
 
-    pure (st', mkShelleyValidatedTx vtx)
-  where
-    innerSt = tickedShelleyLedgerState st
+        st3 :: TickedLedgerState (ShelleyBlock proto era) ValuesMK
+        st3 = unstowLedgerTables st2
+
+        st4 :: TickedLedgerState (ShelleyBlock proto era) TrackingMK
+        st4 = calculateDifferenceTicked st0 st3
```
[the remark this answers is not in the document]
Comment (dnadales) [reply ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Mempool.hs:261]: Yeah, I think so.

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/ShelleyHFC.hs [hunk 17]
```diff
@@ -359,3 +395,88 @@ instance ( ShelleyBasedEra era
...
+
+-- | We use this type for clarity, and because we don't want to declare
+-- 'FromCBOR' and 'ToCBOR' for 'NS'; serializations of sum involves design
+-- decisions that cannot be captured by an all-purpose 'NS' instance
+newtype ShelleyTxOut eras =
+    ShelleyTxOut {unShelleyTxOut :: NS TxOutWrapper eras}
+  deriving (Generic)
+
+-- TODO Can't reuse the 'NS' instance because of its use of 'SOP.Compose', so I
+-- inlined it
```
[the remark this answers is not in the document]
Comment (dnadales) [reply ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/ShelleyHFC.hs:411]: I don't think so, but perhaps we should discuss this with Javier once he's back.

**Review by dnadales — APPROVED** [review]
I'm marking it as approved to signal that I reviewed it. There are minor suggestions that we can incorporate, and for the issues that require more work perhaps we can submit a follow-up PR so that @Jasagredo can review it.

File: ouroboros-consensus-cardano/ouroboros-consensus-cardano.cabal [hunk 18]
```diff
@@ -68,5 +73,35 @@ library
...
+                     , cardano-lmdb
+                     , cardano-slotting
+                     , io-classes
+
+  default-language:    Haskell2010
+  ghc-options:         -Wall
+                       -threaded
+                       -rtsopts
+                       "-with-rtsopts=-T -I0 -N2 -A16m"
+                       -Wno-unticked-promoted-constructors
```
[the remark this answers is not in the document]
Comment (jorisdral) [reply ouroboros-consensus-cardano/ouroboros-consensus-cardano.cabal:107]: See https://github.com/input-output-hk/ouroboros-network/pull/4391#discussion_r1143276272

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Eras.hs [hunk 19]
```diff
@@ -381,3 +389,50 @@ instance ShelleyBasedEra (ConwayEra c) => Core.TranslateEra (ConwayEra c) WrapTx
...
+      . Babbage.translateTxOut
+      . unTxOutWrapper
+
+instance ShelleyBasedEra (ConwayEra c) => Core.TranslateEra (ConwayEra c) TxOutWrapper where
+  type TranslationError (ConwayEra c) TxOutWrapper = Void
+  translateEra _ctxt =
+        pure
+      . TxOutWrapper
+      . Conway.translateTxOut
+      . unTxOutWrapper
```
[the remark this answers is not in the document]
Comment (jorisdral) [reply ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Eras.hs:438]: Done

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Ledger.hs [hunk 20]
```diff
@@ -232,27 +243,205 @@ newtype ShelleyTransition = ShelleyTransitionInfo {
...
+  zipLedgerTables2A
+    f
+    (ShelleyLedgerTables utxoL)
+    (ShelleyLedgerTables utxoC)
+    (ShelleyLedgerTables utxoR) =
+      ShelleyLedgerTables <$> f utxoL utxoC utxoR
+
+  foldLedgerTables f (ShelleyLedgerTables utxo) = f utxo
+
+  foldLedgerTables2 f (ShelleyLedgerTables utxoL) (ShelleyLedgerTables utxoR) = f utxoL utxoR
```
[the remark this answers is not in the document]
Comment (jorisdral) [reply ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Ledger.hs:300]: Done

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Ledger.hs [hunk 21]
```diff
@@ -261,14 +450,16 @@ data instance Ticked (LedgerState (ShelleyBlock proto era)) = TickedShelleyLedge
...
     , tickedShelleyLedgerState      :: !(SL.NewEpochState era)
+    , tickedShelleyLedgerTables     ::
+        !(LedgerTables (LedgerState (ShelleyBlock proto era)) mk)
     }
   deriving (Generic)
 
-deriving instance ShelleyBasedEra era
-               => NoThunks (Ticked (LedgerState (ShelleyBlock proto era)))
+deriving instance (ShelleyBasedEra era
+                  , NoThunks (mk (SL.TxIn (EraCrypto era)) (Core.TxOut era))) => NoThunks (Ticked1 (LedgerState (ShelleyBlock proto era)) mk)
```
[the remark this answers is not in the document]
Comment (jorisdral) [reply ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Ledger.hs:459]: Done

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Ledger.hs [hunk 22]
```diff
@@ -282,17 +473,17 @@ instance ShelleyBasedEra era => IsLedger (LedgerState (ShelleyBlock proto era))
                               , shelleyLedgerState
                               , shelleyLedgerTransition
                               } =
-      swizzle appTick <&> \l' ->
+    swizzle appTick <&> \l' ->
```
[the remark this answers is not in the document]
Comment (jorisdral) [reply ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Ledger.hs:476]: Done

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Ledger.hs [hunk 23]
```diff
@@ -385,7 +582,7 @@ instance Show ShelleyReapplyException where
 
 instance Exception.Exception ShelleyReapplyException where
 
-applyHelper ::
+applyHelper :: forall proto m era.
      (ShelleyCompatible proto era, Monad m)
```
[the remark this answers is not in the document]
Comment (jorisdral) [reply ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Ledger.hs:586]: Done

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Node/TPraos.hs [hunk 24]
```diff
@@ -312,20 +313,23 @@ protocolInfoTPraosShelleyBased ProtocolParamsShelleyBased {
...
+      in
+      ShelleyLedgerState {
         shelleyLedgerTip        = Origin
-      , shelleyLedgerState      =
-          registerGenesisStaking (SL.sgStaking genesis) $
-            SL.initialState genesis additionalGenesisConfig
+      , shelleyLedgerState      = st `withUtxoSL` shelleyUTxOTable emptyLedgerTables
       , shelleyLedgerTransition = ShelleyTransitionInfo {shelleyAfterVoting = 0}
+      , shelleyLedgerTables     = ShelleyLedgerTables $ projectUtxoSL st
       }
```
[the remark this answers is not in the document]
Comment (jorisdral) [reply ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Node/TPraos.hs:326]: Done

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/ShelleyHFC.hs [hunk 25]
```diff
@@ -359,3 +395,88 @@ instance ( ShelleyBasedEra era
         Comp . WrapValidatedGenTx
       . mkShelleyValidatedTx . SL.coerceValidated
     <$> SL.translateValidated @era @WrapTx ctxt (SL.coerceValidated vtx)
+
+{-------------------------------------------------------------------------------
+  A wrapper helpful for Ledger HD
```
[the remark this answers is not in the document]
Comment (jorisdral) [reply ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/ShelleyHFC.hs:400]: done

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/ShelleyHFC.hs [hunk 26]
```diff
@@ -359,3 +395,88 @@ instance ( ShelleyBasedEra era
...
+instance SOP.All ShelleyBasedEra eras => Show (ShelleyTxOut eras) where
+  showsPrec =
+      \p (ShelleyTxOut ns) -> showParen (p > 10) $ showString "ShelleyTxOut " . go ns
+    where
+      go :: SOP.All ShelleyBasedEra eras' => NS TxOutWrapper eras' -> ShowS
+      go = showParen True . \case
+        Z l -> showString "Z " . shows l
+        S r -> showString "S " . go r
+
+-- unline SOP.nsToIndex, this is not restricted to the interval [0, 24)
```
[the remark this answers is not in the document]
Comment (jorisdral) [reply ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/ShelleyHFC.hs:435]: Done

File: ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/CanHardFork.hs (hunk 10 above)
Comment (Jasagredo) [reply ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/CanHardFork.hs:408]: No they are not, they have diffs

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Ledger.hs [hunk 27]
```diff
@@ -418,7 +636,13 @@ applyHelper f cfg blk TickedShelleyLedgerState{
...
         )
 
-    return $ ledgerResult <&> \newNewEpochState -> ShelleyLedgerState {
+    let track ::
+             LedgerState (ShelleyBlock proto era) ValuesMK
+          -> LedgerState (ShelleyBlock proto era) TrackingMK
+        track = calculateDifference stBefore
+
+
+    return $ ledgerResult <&> \newNewEpochState -> forgetLedgerTablesValues $ track $ unstowLedgerTables $ ShelleyLedgerState {
```
[the remark this answers is not in the document]
Comment (Jasagredo) [reply ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Ledger.hs]: This should be very trivial, we don't allocate more stuff than what is already there so ignoring for now

**Pull request merged**